internal/proto/message.go

```diff
@@ -122,139 +114,141 @@ func (m *ClientMessage) SetRetryable(retryable bool) {
-		} else {
-			binary.LittleEndian.PutUint16(bytes[offset+IntSizeInBytes:], currentFrame.flags)
+	lastIndex := len(m.Frames) - 1
+	for i, frame := range m.Frames {
+		binary.LittleEndian.PutUint32(bytes[offset:], uint32(len(frame.Content)+SizeOfFrameLengthAndFlags))
+		flags := frame.flags
+		if i == lastIndex {
+			flags |= IsFinalFlag
 		}
+		binary.LittleEndian.PutUint16(bytes[offset+IntSizeInBytes:], flags)
 		offset += SizeOfFrameLengthAndFlags
-		copy(bytes[offset:], currentFrame.Content)
-		offset += len(currentFrame.Content)
-		currentFrame = currentFrame.next
+		copy(bytes[offset:], frame.Content)
+		offset += len(frame.Content)
 	}
 	return offset
 }
 
 func (m *ClientMessage) DropFragmentationFrame() {
-	m.StartFrame = m.StartFrame.next
+	m.Frames = m.Frames[1:]
+}
+
+func (m *ClientMessage) HasEventFlag() bool {
+	return m.Frames[0].HasEventFlag()
+}
+
+func (m *ClientMessage) HasBackupEventFlag() bool {
+	return m.Frames[0].HasBackupEventFlag()
+}
+
+func (m *ClientMessage) HasFinalFrame() bool {
+	return m.Frames[len(m.Frames)-1].IsFinalFrame()
+}
+
+func (m *ClientMessage) HasUnFragmentedMessageFlags() bool {
+	return m.Frames[0].HasUnFragmentedMessageFlags()
 }
 
 // ForwardFrameIterator
 type ForwardFrameIterator struct {
-	nextFrame *Frame
+	frames []*Frame
+	next   int
 }
 
-func NewForwardFrameIterator(frame *Frame) *ForwardFrameIterator {
-	return &ForwardFrameIterator{frame}
+func NewForwardFrameIterator(frames []*Frame) *ForwardFrameIterator {
+	return &ForwardFrameIterator{frames: frames}
 }
 
-func (forwardFrameIterator *ForwardFrameIterator) Next() *Frame {
-	result := forwardFrameIterator.nextFrame
-	if result != nil {
-		forwardFrameIterator.nextFrame = forwardFrameIterator.nextFrame.next
+func (it *ForwardFrameIterator) Next() *Frame {
+	if it.next >= len(it.frames) {
+		return nil
 	}
+	result := it.frames[it.next]
+	it.next++
 	return result
 }
 
-func (forwardFrameIterator *ForwardFrameIterator) HasNext() bool {
-	return forwardFrameIterator.nextFrame != nil
+func (it *ForwardFrameIterator) HasNext() bool {
+	return it.next < len(it.frames)
 }
 
-func (forwardFrameIterator *ForwardFrameIterator) PeekNext() *Frame {
-	return forwardFrameIterator.nextFrame
+func (it *ForwardFrameIterator) PeekNext() *Frame {
+	return it.frames[it.next]
 }
 
-// Frame
 type Frame struct {
 	Content []byte
 	flags   uint16
-	next    *Frame
 }
 
-// NewFrame create with content
+// NewFrame creates a Frame with content
 func NewFrame(content []byte) *Frame {
 	return &Frame{Content: content, flags: DefaultFlags}
 }
 
-// NewFrame create with content with flags
+// NewFrameWith creates a Frame with content and flags
 func NewFrameWith(content []byte, flags uint16) *Frame {
 	return &Frame{Content: content, flags: flags}
 }
 
 // Copy frame
-func (frame Frame) Copy() *Frame {
-	newFrame := NewFrameWith(frame.Content, frame.flags)
-	newFrame.next = frame.next
-	return newFrame
+func (frame *Frame) Copy() *Frame {
```

Comment:
Copying is not required except the first frame, since I've removed the `.Next` field. That's a placeholder function until I change the protocol generator.

Reply:
Then it probably makes sense to add a TODO here to make the intent more obvious.

Reply:
Right, my bad.

Reply:
Added a TODO at: https://github.com/hazelcast/hazelcast-go-client/pull/555/commits/f8335e3a66165c8383ce93ed08fc6127f3af1375